modules/swagger-codegen/src/main/resources/rust/api.mustache

```diff
@@ -0,0 +1,103 @@
+{{>partial_header}}
+use std::rc::Rc;
+use std::borrow::Borrow;
+
+use hyper;
+use serde_json;
+use futures;
+use futures::{Future, Stream};
+
+use super::{Error, configuration, models};
+
+pub struct {{{classname}}}Impl<C: hyper::client::Connect> {
+    configuration: Rc<configuration::Configuration<C>>,
+}
+
+impl<C: hyper::client::Connect> {{{classname}}}Impl<C> {
+    pub fn new(configuration: Rc<configuration::Configuration<C>>) -> {{{classname}}}Impl<C> {
+        {{{classname}}}Impl {
+            configuration: configuration,
+        }
+    }
+}
+
+pub trait {{classname}} {
+{{#operations}}
+{{#operation}}
+    fn {{{operationId}}}(&self, {{#allParams}}{{paramName}}: {{{dataType}}}{{#hasMore}}, {{/hasMore}}{{/allParams}}) -> Box<Future<Item = {{^returnType}}(){{/returnType}}{{#returnType}}{{{returnType}}}{{/returnType}}, Error = Error>>;
+{{/operation}}
+{{/operations}}
+}
+
+
+impl<C: hyper::client::Connect>{{classname}} for {{classname}}Impl<C> {
+{{#operations}}
+{{#operation}}
+    fn {{{operationId}}}(&self, {{#allParams}}{{paramName}}: {{#isString}}&str{{/isString}}{{^isString}}{{{dataType}}}{{/isString}}{{#hasMore}}, {{/hasMore}}{{/allParams}}) -> Box<Future<Item = {{^vendorExtensions.x-returnTypeInMethod}}(){{/vendorExtensions.x-returnTypeInMethod}}{{#vendorExtensions.x-returnTypeInMethod}}{{{.}}}{{/vendorExtensions.x-returnTypeInMethod}}, Error = Error>> {
+        let configuration: &configuration::Configuration<C> = self.configuration.borrow();
+        let mut req = hyper::Request::new(
+            hyper::Method::{{httpMethod}},
+            format!("{}{{path}}", configuration.base_path{{#allParams}}, {{paramName}}={{paramName}}{{/allParams}}));
```

Comment:
Ok, so, the code would look like:

```rust
        let configuration: &configuration::Configuration<C> = self.configuration.borrow();

        let method = hyper::Method::{{httpMethod}};
        
        // { if empty query args }
        let uri = format!("{}{{path}}", configuration.base_path{{#allParams}}, {{paramName}}={{paramName}}{{/allParams}}));
        // { else }
        let query = url::form_urlencoded::Serializer::new(String::new())
            {{#queryParams}}.append_pair("{{paramName}}", {{paramName}}){{/queryParams}} // only for query params
            .finish();
        let uri = format!("{}{{path}}{}", configuration.base_path, query{{#allParams}}, {{paramName}}={{paramName}}{{/allParams}}));
        // { end }
        
        let mut req = hyper::Request::new(method, uri);
        
        // { if header args }
        let mut headers = req.headers_mut();
        {{#headerParams}}
        headers.set_raw("{{paramName}}", "{{paramValue}}");
        {{/headerParams}}
        // { end }

        // { if body args }
        let serialized = serde_json::to_string(body).unwrap();
        req.headers_mut().set(hyper::header::ContentType::json());
        req.headers_mut().set(hyper::header::ContentLength(serialized.len() as u64));
        req.set_body(serialized);
        // { end }

        // send request
        Box::new(
            configuration.client.request(req).and_then(|res| { res.body().concat2() })
            .map_err(|e| Error::from(e))
            // { if no response expected }
            .and_then(|_| futures::future::ok(()))
            // { else }
            .and_then(|body| {
                let parsed: Result<{{returnType}}, _> = serde_json::from_slice(&body);
                parsed.map_err(|e| Error::from(e))
            }).map_err(|e| Error::from(e))
            // { end}
        )
```

This covers path, query, header params. Trying to figure where do form params land and there's also the "file" thing that requires multipart upload.

As on extracting the logic, I think it's better to have version 1 with code explicit. There are a few options on how to extract the logic, but I'm not sure which one would be more effective in terms of calls/memory allocation and still researching that.